geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/KryoLazyDensityFilter.java

```diff
@@ -0,0 +1,359 @@
+    }
+
+    private static String getGeomField() {
+        GeometryDescriptor gd = sft.getGeometryDescriptor();
+        if (gd == null) {
+            return null;
+        } else {
+            return gd.getLocalName();
+        }
+    }
+
+    private static int getGeomIndex() {
+        return sft.indexOf(getGeomField());
+    }
+
+    private static Envelope getDensityEnvelope(Hints hints) {
+        Envelope e = (Envelope) hints.get(QueryHints.DENSITY_BBOX());
+        return e;
+    }
+
+    private static int getDensityWidth(Hints hints) {
+        int width = (int) hints.get(QueryHints.DENSITY_WIDTH());
+        return width;
+    }
+
+    private static int getDensityHeight(Hints hints) {
+        int height = (int) hints.get(QueryHints.DENSITY_HEIGHT());
+        return height;
+    }
+
+    private static String getDensityWeight(Hints hints) {
+        String weight = (String) hints.get(QueryHints.DENSITY_WEIGHT());
+        return weight;
+    }
+
+
+    /**
+     * Gets the weight for a feature from a double attribute
+     */
+    private double getWeightFromDouble(int i, SimpleFeature sf) {
+        Double d = (Double) sf.getAttribute(i);
+        if (d == null) {
+            return 0.0;
+        } else {
+            return d;
+        }
+    }
+
+    /**
+     * Tries to convert a non-double attribute into a double
+     */
+    private double getWeightFromNonDouble(int i, SimpleFeature sf) {
+        Object d = sf.getAttribute(i);
+        if (d == null) {
+            return 0.0;
+        } else {
+            Double converted = Converters.convert(d, Double.class);
+            if (converted == null) {
+                return 1.0;
+            } else {
+                return converted;
+            }
+        }
+    }
+
+    /**
+     * Evaluates an arbitrary expression against the simple feature to return a weight
+     */
+    private double getWeightFromExpression(Expression e, SimpleFeature sf) {
+        Double d = e.evaluate(sf, Double.class);
+        if (d == null) {
+            return 0.0;
+        } else {
+            return d;
+        }
+    }
+
+    /*
+     * Returns the weight of a simple feature
+     */
+    double weightFn(SimpleFeature sf) {
+        double returnVal = 0.0;
+        if (weightIndex == -2) {
+            returnVal = 1.0;
+        } else if (weightIndex == -1) {
+            try {
+                Expression expression = ECQL.toExpression((String) options.get(WEIGHT_OPT));
+                returnVal = getWeightFromExpression(expression, sf);
+            } catch (Exception e) {
+                System.out.println("Exception");
+            }
+        } else if (sft.getDescriptor(weightIndex).getType().getBinding() == Double.class) {
+            returnVal = getWeightFromDouble(weightIndex, sf);
+        } else {
+            returnVal = getWeightFromNonDouble(weightIndex, sf);
+        }
+        return returnVal;
+    }
+
+    boolean isPoints() {
```

Comment:
there is so much code in this file that is the same as the accumulo implementation - we should extract out the common details (probably not for this PR)